File: shell_wrapper/single_thread_hkdf.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "shell_wrapper/single_thread_hkdf.h"

#include <sys/types.h>

#include <cstddef>
#include <cstdint>
#include <memory>
#include <string>
#include <utility>

#include "absl/status/status.h"
#include "absl/strings/string_view.h"
#include "include/cxx.h"
#include "shell_encryption/prng/single_thread_hkdf_prng.h"
#include "shell_wrapper/shell_types.h"
#include "shell_wrapper/single_thread_hkdf.rs.h"
#include "tink/subtle/common_enums.h"
#include "tink/subtle/hkdf.h"

using secure_aggregation::MakeFfiStatus;

FfiStatus GenerateSingleThreadHkdfSeed(std::unique_ptr<std::string>& out) {
  auto statusor = rlwe::SingleThreadHkdfPrng::GenerateSeed();
  if (!statusor.ok()) {
    return MakeFfiStatus(std::move(statusor.status()));
  };
  out = std::make_unique<std::string>(*std::move(statusor));
  return MakeFfiStatus();
}
FfiStatus CreateSingleThreadHkdf(rust::Slice<const uint8_t> seed,
                                 SingleThreadHkdfWrapper& out) {
  auto statusor = rlwe::SingleThreadHkdfPrng::Create(ToAbslStringView(seed));
  if (!statusor.ok()) {
    return MakeFfiStatus(std::move(statusor).status());
  }
  out.ptr = *std::move(statusor);
  return MakeFfiStatus();
}

FfiStatus Rand8(SingleThreadHkdfWrapper& prng, uint8_t& out) {
  if (prng.ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  auto statusor = prng.ptr->Rand8();
  if (!statusor.ok()) {
    return MakeFfiStatus(std::move(statusor).status());
  }
  out = *statusor;
  return MakeFfiStatus();
}

size_t SingleThreadHkdfSeedLength() {
  return rlwe::SingleThreadHkdfPrng::SeedLength();
}
FfiStatus ComputeHkdfWrapper(rust::Slice<const uint8_t> input,
                             rust::Slice<const uint8_t> salt,
                             rust::Slice<const uint8_t> info, size_t out_len,
                             std::unique_ptr<std::string>& out) {
  auto statusor = crypto::tink::subtle::Hkdf::ComputeHkdf(
      crypto::tink::subtle::SHA256, ToAbslStringView(input),
      ToAbslStringView(salt), ToAbslStringView(info), out_len);
  if (!statusor.ok()) {
    return MakeFfiStatus(std::move(statusor).status());
  }
  out = std::make_unique<std::string>(*std::move(statusor));
  return MakeFfiStatus();
}

File: shell_wrapper/single_thread_hkdf.rs
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Rust wrapper around https://github.com/google/shell-encryption/blob/master/shell_encryption/prng/single_thread_hkdf_prng.h

use status::rust_status_from_cpp;

#[cxx::bridge]
mod ffi {
    struct SingleThreadHkdfWrapper {
        ptr: UniquePtr<SingleThreadHkdfPrng>,
    }

    unsafe extern "C++" {
        type FfiStatus = shell_types::ffi::FfiStatus;

        include!("shell_encryption/prng/single_thread_hkdf_prng.h");
        #[namespace = "rlwe"]
        type SingleThreadHkdfPrng;

        include!("shell_wrapper/single_thread_hkdf.h");
        pub fn GenerateSingleThreadHkdfSeed(out: &mut UniquePtr<CxxString>) -> FfiStatus;
        pub fn CreateSingleThreadHkdf(seed: &[u8], out: &mut SingleThreadHkdfWrapper) -> FfiStatus;
        pub fn Rand8(prng: &mut SingleThreadHkdfWrapper, out: &mut u8) -> FfiStatus;

        pub fn SingleThreadHkdfSeedLength() -> usize;

        pub fn ComputeHkdfWrapper(
            input: &[u8],
            salt: &[u8],
            info: &[u8],
            out_len: usize,
            out: &mut UniquePtr<CxxString>,
        ) -> FfiStatus;
    }
}

use shell_types::ffi::CloneString;
use shell_types::ffi::EmptyString;

/// Contains a pointer to a PRNG. Re-exported from cxx.
pub use ffi::SingleThreadHkdfWrapper;
pub struct SeedWrapper(cxx::UniquePtr<cxx::CxxString>);
impl Clone for SeedWrapper {
    fn clone(&self) -> Self {
        if self.0.is_null() {
            Self(cxx::UniquePtr::null())
        } else {
            Self(CloneString(&self.0))
        }
    }
}

impl std::ops::Deref for SeedWrapper {
    type Target = cxx::CxxString;
    fn deref(&self) -> &cxx::CxxString {
        if self.0.is_null() {
            EmptyString()
        } else {
            &self.0
        }
    }
}

pub use ffi::SingleThreadHkdfSeedLength as seed_length;

/// Generates a valid seed for the Prng.
///
/// Fails on internal cryptographic errors.
pub fn generate_seed() -> Result<SeedWrapper, status::StatusError> {
    let mut out = cxx::UniquePtr::null();
    rust_status_from_cpp(ffi::GenerateSingleThreadHkdfSeed(&mut out))?;
    Ok(SeedWrapper(out))
}

/// Constructs a new PRNG from a seed. See
/// https://github.com/google/shell-encryption/blob/master/shell_encryption/prng/single_thread_hkdf_prng.h
///
/// Fails if the key is not the expected size or on internal cryptographic
/// errors.
pub fn create(seed: &SeedWrapper) -> Result<SingleThreadHkdfWrapper, status::StatusError> {
    let mut out = ffi::SingleThreadHkdfWrapper { ptr: cxx::UniquePtr::null() };
    rust_status_from_cpp(ffi::CreateSingleThreadHkdf(seed.as_bytes(), &mut out))?;
    Ok(out)
}

/// Returns a random byte.
///
/// Fails on internal cryptographic errors.
pub fn rand8(prng: &mut SingleThreadHkdfWrapper) -> Result<u8, status::StatusError> {
    // SAFETY: Rand8 advances the state of the C++ PRNG contained in `prng`, and
    // returns one u8 that does not keep any references). No lifetime constraints.
    let mut out: u8 = 0;
    rust_status_from_cpp(ffi::Rand8(prng, &mut out))?;
    Ok(out)
}

pub fn compute_hkdf(
    input: &[u8],
    salt: &[u8],
    info: &[u8],
    out_len: usize,
) -> Result<SeedWrapper, status::StatusError> {
    let mut out = cxx::UniquePtr::null();
    let status = ffi::ComputeHkdfWrapper(input.into(), salt.into(), info.into(), out_len, &mut out);
    rust_status_from_cpp(status)?;
    Ok(SeedWrapper(out))
}
